les/peer.go

```diff
@@ -1050,49 +1063,86 @@ func (ps *clientPeerSet) unSubscribe(sub clientPeerSubscriber) {
 
 // register adds a new peer into the peer set, or returns an error if the
 // peer is already known.
-func (ps *clientPeerSet) register(peer *clientPeer) error {
+func (ps *clientPeerSet) register(p *clientPeer) error {
 	ps.lock.Lock()
-	defer ps.lock.Unlock()
-
 	if ps.closed {
+		ps.lock.Unlock()
 		return errClosed
 	}
-	if _, exist := ps.peers[peer.id]; exist {
+	if _, ok := ps.active[p.id]; ok {
+		ps.lock.Unlock()
 		return errAlreadyRegistered
 	}
-	ps.peers[peer.id] = peer
-	for _, sub := range ps.subscribers {
-		sub.registerPeer(peer)
+	delete(ps.inactive, p.id)
+	ps.active[p.id] = p
+
+	peers := make([]clientPeerSubscriber, len(ps.subscribers))
+	copy(peers, ps.subscribers)
+	ps.lock.Unlock()
+
+	for _, n := range peers {
+		n.registerPeer(p)
 	}
 	return nil
 }
 
 // unregister removes a remote peer from the peer set, disabling any further
 // actions to/from that particular entity. It also initiates disconnection
 // at the networking layer.
-func (ps *clientPeerSet) unregister(id string) error {
+func (ps *clientPeerSet) unregister(p *clientPeer) error {
 	ps.lock.Lock()
-	defer ps.lock.Unlock()
+	if _, ok := ps.active[p.id]; !ok {
+		ps.lock.Unlock()
+		return errNotRegistered
+	} else {
+		delete(ps.active, p.id)
+		ps.inactive[p.id] = p
+		peers := make([]clientPeerSubscriber, len(ps.subscribers))
+		copy(peers, ps.subscribers)
+		ps.lock.Unlock()
+
+		for _, n := range peers {
+			n.unregisterPeer(p)
+		}
+		return nil
+	}
+}
 
-	p, ok := ps.peers[id]
-	if !ok {
+// disconnect removes a remote peer from either the active or inactive set and
+// initiates disconnection at the networking layer.
+func (ps *clientPeerSet) disconnect(id string) error {
```

Comment:
Why do we really need this logic? I don't think we should classify `active` and `inactive` status for clients.

In the LES, server can always send the message to clients(e.g. header broadcast) but the inactive client is not allowed to send more requests if it's idle.

The `inactive` notion is only necessary in the clientpool right?